include/deal.II/lac/solver_gmres.h

```diff
@@ -710,77 +722,202 @@ SolverGMRES<VectorType>::givens_rotation(Vector<double> &h,
+     * the procedure every fifth step and sets the flag to true in that case.
+     * All subsequent iterations use re-orthogonalization.
+     * Calls the signal re_orthogonalize_signal if it is connected.
+     */
+    template <class VectorType>
+    inline double
+    iterated_gram_schmidt(
+      const typename SolverGMRES<VectorType>::AdditionalData::
+        OrthogonalizationStrategy orthogonalization_strategy,
+      const internal::SolverGMRESImplementation::TmpVectors<VectorType>
+        &                                       orthogonal_vectors,
+      const unsigned int                        dim,
+      const unsigned int                        accumulated_iterations,
+      VectorType &                              vv,
+      Vector<double> &                          h,
+      bool &                                    reorthogonalize,
+      const boost::signals2::signal<void(int)> &reorthogonalize_signal =
+        boost::signals2::signal<void(int)>())
     {
-      double htmp = vv * orthogonal_vectors[0];
-      h(0) += htmp;
-      for (unsigned int i = 1; i < dim; ++i)
+      Assert(dim > 0, ExcInternalError());
+      const unsigned int inner_iteration = dim - 1;
+
+      // need initial norm for detection of re-orthogonalization, see below
+      double     norm_vv_start = 0;
+      const bool consider_reorthogonalize =
+        (reorthogonalize == false) && (inner_iteration % 5 == 4);
+      if (consider_reorthogonalize)
+        norm_vv_start = vv.l2_norm();
+
+      for (unsigned int i = 0; i < dim; ++i)
+        h[i] = 0;
+
+      for (unsigned int c = 0; c < 2;
+           ++c) // 0: orthogonalize, 1: reorthogonalize
         {
-          htmp = vv.add_and_dot(-htmp,
-                                orthogonal_vectors[i - 1],
-                                orthogonal_vectors[i]);
-          h(i) += htmp;
+          // Orthogonalization
+          double norm_vv = 0.0;
+
+          if (orthogonalization_strategy ==
+              SolverGMRES<VectorType>::AdditionalData::
+                OrthogonalizationStrategy::modified_gram_schmidt)
+            {
+              double htmp = vv * orthogonal_vectors[0];
+              h(0) += htmp;
+              for (unsigned int i = 1; i < dim; ++i)
+                {
+                  htmp = vv.add_and_dot(-htmp,
+                                        orthogonal_vectors[i - 1],
+                                        orthogonal_vectors[i]);
+                  h(i) += htmp;
+                }
+
+              norm_vv = std::sqrt(
+                vv.add_and_dot(-htmp, orthogonal_vectors[dim - 1], vv));
+            }
+          else if (orthogonalization_strategy ==
+                   SolverGMRES<VectorType>::AdditionalData::
+                     OrthogonalizationStrategy::classical_gram_schmidt)
+            {
+              dot_and_add(dim, vv, orthogonal_vectors, h);
+              norm_vv = substract_and_norm(dim, orthogonal_vectors, h, vv);
+            }
```

Comment:
@kronbichler This is how the unified code looks like. What do you think? I still need to work on the vectorization.